optical/converter/tfrecord.py

```diff
@@ -0,0 +1,125 @@
+"""
+__author__: HashTagML
+license: MIT
+Created: Friday, 16th April 2021
+"""
+
+import os
+import io
+import warnings
+from typing import Union
+from pathlib import Path
+
+from PIL import Image
+import numpy as np
+import pandas as pd
+
+try:
+    import tensorflow as tf
+except Exception:
+    tf = None
+
+from .base import FormatSpec
+from .utils import tf_parse_example
+
+
+class Tfrecord(FormatSpec):
+    def __init__(self, root: Union[str, os.PathLike]):
+        self.root = root
+        self._has_image_split = False
+        image_path = Path(root) / "images"
+        image_path.mkdir(parents=True, exist_ok=True)
+        self._annotation_dir = root
+        self._image_dir = image_path
+        self._splits = self._find_splits()
+        if tf:
+            self._resolve_dataframe()
+        else:
+            warnings.warn("Please install tensorflow for support of tfrecord")
+
+    def _find_splits(self):
+        ann_splits = [x.stem for x in Path(self.root).glob("*.tfrecord")]
+
+        assert ann_splits, "directory doesn't have tfrecords"
+        if len(ann_splits) > 1:
+            self._has_image_split = True
+
+        return ann_splits
+
+    def _resolve_dataframe(self):
+        img_filenames = []
+        img_widths = []
+        img_heights = []
+        cls_names = []
+        x_mins = []
+        y_mins = []
+        box_widths = []
+        box_heights = []
+        splits = []
+        cls_ids = []
+        for split in self._splits:
+            tf_record = str(Path(self.root) / f"{split}.tfrecord")
+            render = tf.data.TFRecordDataset(tf_record)
+            dataset = render.map(tf_parse_example)
+            img_dir = Path(self.root) / "images" / split
+            img_dir.mkdir(parents=True, exist_ok=True)
+            for data in dataset:
```

Comment:
I think since it involves image io operations. we can parallelize this loop to make it faster.

Reply:
Parallelized the image saving

Reply:
@satheeshkatipomu @bharatkumarreddy joblib is so weird. Sometimes it's just much slower than a simple python for loop, especially for a bunch of small tasks. On the other hand it's blazing fast. Can we please have a comparison here?

Reply:
Yes, parallelization has some overhead, so will be slower if our individual tasks take very less time , but works better when tasks take longer. Better to compare them

@bharatkumarreddy , we can move that for loop as a separate method, can parallelize whole decoding part